device identity

piotrbartman opened the pull request:
implements QubesOS/qubes-issues/issues/9325

piotrbartman committed:
q-dev: port

piotrbartman committed:
q-dev: check identity

piotrbartman committed:
q-dev: refactor device_protocol.py

piotrbartman committed:
q-dev: implements device_id

piotrbartman committed:
q-dev: fix detaching

piotrbartman committed:
q-dev: update device utils

piotrbartman committed:
q-dev: assignment.devices

piotrbartman committed:
q-dev: backward compatibility

piotrbartman committed:
q-dev: fix detaching

piotrbartman committed:
q-dev: auto-attaching order

piotrbartman committed:
q-dev: fix auto-attach

piotrbartman committed:
q-dev: add more tests

piotrbartman committed:
q-dev: update utils

piotrbartman committed:
q-dev: correct english grammar

piotrbartman committed:
q-dev: async confirmation

piotrbartman committed:
q-dev: update utils.py

piotrbartman committed:
q-dev: forbid usb assignment options

piotrbartman committed:
q-dev: updated tests and wait for attachment to be done

piotrbartman committed:
q-dev: rename attach-confirm -> qubes-device-attach-confirm

piotrbartman committed:
q-dev: update tests

piotrbartman committed:
typo

piotrbartman committed:
q-dev: cleanup

piotrbartman committed:
q-dev: call attach-confirm socket directly

piotrbartman committed:
q-dev: update common part with admin-core

marmarek commented:
from openQA:
```
Traceback (most recent call last):
  File "/usr/lib64/python3.13/contextlib.py", line 85, in inner
    return func(*args, **kwds)
  File "/usr/lib/python3.13/site-packages/qubesusbproxy/tests.py", line 490, in test_010_assign
    self.frontend.start()
    ~~~~~~~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/qubes/tests/extra.py", line 76, in start
    return self._loop.run_until_complete(
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        self._vm.start(start_guid=start_guid))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.13/asyncio/base_events.py", line 721, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/usr/lib/python3.13/site-packages/qubes/vm/qubesvm.py", line 1287, in start
    await self.fire_event_async('domain-start',
                                start_guid=start_guid)
  File "/usr/lib/python3.13/site-packages/qubes/events.py", line 234, in fire_event_async
    effect = task.result()
  File "/usr/lib/python3.13/site-packages/qubesusbproxy/core3ext.py", line 761, in on_domain_start
    asyncio.ensure_future(self.attach_and_notify(vm, assignment))
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.13/asyncio/tasks.py", line 742, in ensure_future
    raise TypeError('An asyncio.Future, a coroutine or an awaitable '
                    'is required')
TypeError: An asyncio.Future, a coroutine or an awaitable is required
```

marmarek commented:
The above error is still the case in the most recent run

marmarek commented:
`qvm-usb ls` reports no devices for me, and also I don't see any errors in the log... I'm not sure if it's related to the above message, but may be separate issue (the above looks to be about attaching, not listing).

fepitre committed:
Use Port for device and make mypy happy

Related to work being done in https://github.com/QubesOS/qubes-core-admin/pull/627

fepitre committed:
Remove core2 tests

fepitre committed:
tests: use AsyncMock

marmarek commented on qubesusbproxy/tests.py:
```diff
@@ -524,76 +667,521 @@ def test_061_auto_attach_on_reconnect(self):
+            qubesusbproxy.core3ext.USBDevice(back, "1-1")
+        )
+
+        self.ext.attach_and_notify = Mock()
```
You can't do that, use patch instead.
It doesn't looks like it, bue extension object actually is a singleton, and this will affect other tests in this process too.
And also, this should be AsyncMock (@fepitre already fixed it in #43).

piotrbartman replied:
> extension object actually is a singleton

Ok, that explains a lot

marmarek replied:
I haven't tried manually fixing this, but if I skip this test class, the other one passes in full, no more "Device connection failed" errors etc :)

piotrbartman committed:
q-dev: replace overriding with patching

marmarek commented on qubesusbproxy/tests.py:
```diff
@@ -722,12 +722,13 @@ def test_010_on_qdb_change_multiple_assignments_including_full(self):
-            self.ext.attach_and_notify.call_args[0][1].options, {"pid": "did"}
-        )
+        with (mock.patch.object(self.ext, "attach_and_notify")
+              as attach_and_notify):
```
I think you need to mock it with `AsyncMock()` (just add as the third arg to patch.object), otherwise you'll get back to the TypeError.

piotrbartman replied:
unittest mock automatically uses `AsyncMock` for coroutines

marmarek replied:
Ah, nice :)

marmarek reviewed (changes requested):
If it's applicable to all tests, maybe it's cleaner to apply patch in `setUp()` (and unpatch in `tearDown()`)? That means calling start() and stop() on the patch object instead of using as a context manager.
But, I see the first test doesn't patch this, so I'm not sure if that's really applicable...

marmarek commented:
This also fails horribly on R4.2 (utils.py tries to import device_protocol, then uses `DeviceCollection.get_assigned_devices()` and probably few more.
Maybe it's time to abandon using the same branch for R4.2 and R4.3?

marmarek commented:
Ok, the last test run on R4.3 shows no regressions. I'll split the branch for R4.2 and then (finally) merge this all.
